mesonbuild/options.py

```diff
@@ -324,7 +332,10 @@ def __post_init__(self, value_: int) -> None:
             choices.append(f'>= {self.min_value!s}')
         if self.max_value is not None:
             choices.append(f'<= {self.max_value!s}')
-        self.choices = ', '.join(choices)
+        self.__choices: str = ', '.join(choices)
+
+    def printable_choices(self) -> T.Optional[T.List[str]]:
+        return [self.__choices]
```

Comment:
Overloading `printable_choices()` to return a description of the valid range is a bit weird. The use of `self.__choices` makes it even less understandable. Would it make sense to move the computation of the string returned by `printable_choices()` into the method itself? I don't think that this code path is performance sensitive and requires caching the computation, but if it does, decorating the method with `@functools.cache` may be a better solution.

Reply:
This is super important actually, because choices needs to be a `Container[T]`, matching the type of the `UserOption[T]`, but printable choices needs to be `List[str] | None`, regardless of what `T` is. The entire point of `printable_choices` is to fix the fact that we both need a string definable set of choices (like those used in Combo and Array options that are essentially a set of valid values, and the choices of integer options, where the constraints are a range of valid values. The status quo is that `IntegerOption` shoves `'>=0, < 100'` into `choices` and acts like it's cool, but then `BooleanOption` puts `[True, False]` in there.